Sources/BaggageContext/Context.swift

```diff
@@ -0,0 +1,287 @@
+///
+/// This allows frameworks and library authors to offer APIs which compose more easily.
+/// Please refer to the "Reference Implementation" notes on each of the requirements to know how to implement this protocol correctly.
+///
+/// ### Relation to `BaggageProtocol`
+/// The `ContextProtocol` does NOT extend `BaggageProtocol`.
+///
+/// A custom context implementation however `MAY` extend both `ContextProtocol` and `BaggageProtocol`.
+/// Doing so enables a custom context to obtain the baggage protocol's capability of getting *and setting*
+/// properties on a baggage directly, like so: `myContext.traceID = "111-222"` without having to type out the baggage
+/// explicitly, as would normally be the case (`myContext.baggage.traceID = "111-222"`).
+///
+/// This difference in capabilities is purposefully designed as such, in order to allow context objects implement `ContextProtocol`
+/// without having to force them to allow _modifying_ the baggage. Some frameworks or libraries may choose to create immutable
+/// context objects, including the baggage they offer there. Users, when wanting to add more contextual information need then to
+/// perform the `asBaggageContext`, mutate the resulting context OR use the `with...` fluent-API functions to obtain a mutated context.
+///
+/// Please note that if a context does conform to `BaggageProtocol` as well, it will most-likely also offer all properties that
+/// users have provided via extensions on it - so be watchful of potential naming conflicts on variable names.
+public protocol ContextProtocol {
+    /// Get the `Baggage` container.
+    var baggage: Baggage { get }
+
+    /// The `Logger` associated with this context carrier.
+    ///
+    /// It automatically populates the loggers metadata based on the `BaggageContext` associated with this context object.
+    ///
+    /// ### Implementation note
+    ///
+    /// Libraries and/or frameworks which conform to this protocol with their "Framework Context" types,
+    /// SHOULD implement this logger by wrapping the "raw" logger associated with  `_logger.with(self.baggage)` function,
+    /// which efficiently handles the bridging of baggage to logging metadata values.
+    ///
+    /// ### Reference Implementation
+    ///
+    /// Writes to the `logger` metadata SHOULD NOT be reflected in the `baggage`,
+    /// however writes to the underlying `baggage` SHOULD be reflected in the `logger`.
+    ///
+    ///     struct MyFrameworkContext: ContextProtocol {
+    ///       var baggage: Baggage
+    ///       private let _logger: Logger
+    ///
+    ///       var logger: Logger {
+    ///         return self._logger.with(self.baggage)
+    ///       }
+    ///     }
+    var logger: Logger { get }
+
+    /// Get a baggage `Context` struct, regardless of the representation that this context protocol is implemented by.
+    var asBaggageContext: Context { get }
```

Comment:
This is an escape hatch to allow users to always to back to the "safe, known, shared" type when they don't know what context to pass.

Reply:
this is confusing to me, we are going from the abstract protocol to the concrete default impl?

Reply:
I think so, in cases where you need to pass it to generic methods?

I'm still a little confused as to if the protocol or the concrete type is what is encouraged to use by API authors in declarations. If the protocol, should we be encouraging generics or existentials?

Reply:
> I'm still a little confused as to if the protocol or the concrete type is what is encouraged to use by API authors in declarations.

We'll rename it to `struct DefaultContext` that should clarify things -- always accept a `Context` and the struct exists if you dont have a context and need to create one.

> If the protocol, should we be encouraging generics or existentials?

Existentials just because the API shapes quickly become insane otherwise, so as all examples are showing: `func hello(who: String, context: Context)`

Reply:
I'll remove the `asBaggageContext` though and replace with a `struct DefaultContext { init(_ Context)` instead

Reply:
This is resolved thanks to introducing get/set on the baggage/logger, we dont need the escape hatch anymore.